pf/internal/defaults.go

```diff
@@ -1,42 +1,139 @@
+	fieldSchema shim.Schema,
+	defaultInfo *tfbridge.DefaultInfo,
+) (resource.PropertyValue, bool, error) {
+	na := resource.NewNullProperty()
+
+	// TODO handle defaultInfo.Config
+	if defaultInfo == nil {
+		return na, false, nil
 	}
-	if d.From != nil {
-		return d.From(res)
-	} else if d.EnvVars != nil {
-		for _, n := range d.EnvVars {
-			if v, ok := os.LookupEnv(n); ok {
-				return v, nil
+	if defaultInfo.From != nil {
+		raw, err := defaultInfo.From(res)
+		if err != nil {
+			return na, false, err
+		}
+		if raw == nil {
+			return na, false, nil
+		}
+		return recoverDefaultValue(raw), true, nil
+
+	} else if defaultInfo.EnvVars != nil {
+		for _, n := range defaultInfo.EnvVars {
+			if str, ok := os.LookupEnv(n); ok {
+				// TODO what is the behavior of an env var set to an emtpy string? Is this the same as
+				// unset, or the same as actually setting to empty?
+				v, err := parseValueFromEnv(fieldSchema, str)
+				return v, true, err
 			}
 		}
 	}
-	return d.Value, nil
+	if defaultInfo.Value != nil {
+		return recoverDefaultValue(defaultInfo.Value), true, nil
+	}
+	return na, false, nil
 }
 
-func SetDefaultValues(res *tfbridge.PulumiResource, fields map[string]*tfbridge.SchemaInfo) error {
-	for key, fld := range fields {
-		if _, ok := res.Properties[resource.PropertyKey(key)]; !ok && fld.Default != nil {
-			// using default value for empty property
-			v, err := getDefaultValue(fld.Default, res)
-			if err != nil {
-				return err
+func parseValueFromEnv(sch shim.Schema, str string) (resource.PropertyValue, error) {
+	var err error
+	switch sch.Type() {
+	case shim.TypeBool:
+		v := false
+		if str != "" {
+			if v, err = strconv.ParseBool(str); err != nil {
+				return resource.NewNullProperty(), err
+			}
+		}
+		return resource.NewBoolProperty(v), nil
+	case shim.TypeInt:
+		v := int(0)
+		if str != "" {
+			iv, iverr := strconv.ParseInt(str, 0, 0)
+			if iverr != nil {
+				return resource.NewNullProperty(), iverr
 			}
-			if pval, ok := v.(resource.PropertyValue); ok {
-				res.Properties[resource.PropertyKey(key)] = pval
+			v = int(iv)
+		}
+		return resource.NewNumberProperty(float64(v)), nil
+	case shim.TypeFloat:
+		v := float64(0.0)
+		if str != "" {
+			if v, err = strconv.ParseFloat(str, 64); err != nil {
+				return resource.NewNullProperty(), err
 			}
-			res.Properties[resource.PropertyKey(key)] = resource.NewPropertyValue(v)
 		}
+		return resource.NewNumberProperty(float64(v)), nil
+	case shim.TypeString:
+		return resource.NewStringProperty(str), nil
+	default:
+		return resource.NewNullProperty(), fmt.Errorf("unknown type for default value: %v", sch.Type())
+	}
+}
+
+func recoverDefaultValue(defaultValue any) resource.PropertyValue {
+	if pv, alreadyPV := defaultValue.(resource.PropertyValue); alreadyPV {
+		return pv
 	}
+	return resource.NewPropertyValue(defaultValue)
+}
+
+// TODO this needs to handle the nested case, setting defaults for properties of object-typed properties.
```

Comment:
I'm working on some combinators here to make it easy to handle the recursive case, in a separate PR.

Reply:
You wouldn't be able to do it correctly so quickly because the codebase doesn't make it obvious how to deal with maxItems=1 flattening and Pulumi-terraform name mangling. I've made progress on https://github.com/pulumi/pulumi-terraform-bridge/pull/1065 and I'll try to get it in piecemeal shortly then we can use some of the same combinators to solve this as well. For example we can walk PropertyValue keeping track of PropertyPath and then use combinators from that PR to translate that to SchemaPath and lookup corresponding schema information.